sdk/metric/view/view.go

```diff
@@ -17,22 +17,197 @@
 
 package view // import "go.opentelemetry.io/otel/sdk/metric/view"
 
-// Config contains configuration options for a view.
+import (
+	"fmt"
+	"regexp"
+
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/sdk/instrumentation"
+	"go.opentelemetry.io/otel/sdk/metric/internal/instrument"
+)
+
+// Config contains Configuration options for a view. An empty config will match
+// all instruments, and do no modifications.
 type Config struct {
-	// TODO (#2837): implement.
+	instrumentName       string
+	instrumentNameRegexp *regexp.Regexp
+	library              instrumentation.Library
+
+	filterKeys  map[attribute.Key]struct{}
+	viewName    string
+	description string
+	// TODO: Aggregation selection
+}
+
+// New returns a new configured view Config.
+func New(opts ...Option) (Config, error) {
+	cfg := Config{}
+
+	for _, opt := range opts {
+		cfg = opt.apply(cfg)
+	}
+
+	if cfg.instrumentName != "" && cfg.instrumentNameRegexp != nil {
+		return Config{}, fmt.Errorf("can not use name selector and name regex in the same view")
+	}
+	if cfg.instrumentNameRegexp != nil && cfg.viewName != "" {
+		return Config{}, fmt.Errorf("can not use name regex with a view name")
+	}
+
+	return cfg, nil
+}
+
+// TransformInstrument will check if an instrument description matches this view
+// and will convert it if it does.
+func (cfg Config) TransformInstrument(desc instrument.Description) (outputDesc instrument.Description, match bool) {
+	if !cfg.matchName(desc.InstrumentName) ||
+		!cfg.matchLibraryName(desc.Library.Name) ||
+		!cfg.matchLibrarySchemaURL(desc.Library.SchemaURL) ||
+		!cfg.matchLibraryVersion(desc.Library.Version) {
+		return instrument.Description{}, false
+	}
+	if cfg.viewName != "" {
+		desc.InstrumentName = cfg.viewName
+	}
+	if cfg.description != "" {
+		desc.InstrumentDescription = cfg.description
+	}
+	return desc, true
+}
+
+// TransformAttributes filters an attribute set to the keys in the config. If no
+// filter was provided the original set is returned.
+func (cfg Config) TransformAttributes(input attribute.Set) attribute.Set {
+	if len(cfg.filterKeys) == 0 {
+		return input
+	}
+	out, _ := input.Filter(attribute.Filter(func(kv attribute.KeyValue) bool {
+		_, ok := cfg.filterKeys[kv.Key]
+		return ok
+	}))
+	return out
+}
+
+// TODO: Provide Transform* for AggregationKind (#2816)
+
+func (cfg Config) matchName(name string) bool {
+	if cfg.instrumentNameRegexp != nil {
+		return cfg.instrumentNameRegexp.MatchString(name)
+	}
+	return cfg.instrumentName == "" || name == cfg.instrumentName
+}
```

Comment:
We may want to unify on a single instrument name matching option that compiles that name as a wildcard containing string. It will unify the options and remove the error that is returned for the overlapping regexp and string names currently, but it will likely require us to write our own wildcard parsing logic (shouldn't be to hard on first look).

Reply:
You can do wildcard matches by using the Regexp, the spec says you have to have the special case if you don't have any other method of doing ranged matches.  Unless this is a blocking issue I would rather not have to try and reimplement string matching that is sort of like regular expressions if I don't have to.

We have to also check for the case when you use a range match with a rename that is invalid.  

I think what might be a solution is to have `MatchName()` clear `instrumentNameRegexp`, and `MatchNameRegex()` clear `instrumentName`?

Reply:
> Unless this is a blocking issue ...

Ah, yes, this is a blocking issue as this currently does not conform to the specification. I should have done a better job communicating this in the review status I provided.

It does not implement wildcard matching: https://go.dev/play/p/MSqIE0d_k2A

```go
package main

import (
	"regexp"
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func TestWildcard(t *testing.T) {
	tests := []struct {
		name    string
		matches []string
	}{
		{name: "*", matches: []string{"every", "inst", "name"}},
		{name: "*temp", matches: []string{"house-temp", "coretemp", "0-temp"}},
		{name: "us1-*", matches: []string{"us1-capacity", "us1-load"}},
		{name: "reader-?", matches: []string{"reader-1", "reader-"}},
		{name: "mem?-*", matches: []string{"mem-size", "mem0-cache"}},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			re, err := regexp.Compile(tt.name)
			require.NoError(t, err)
			for _, m := range tt.matches {
				match := re.FindString(m)
				assert.Equal(t, m, match)
			}
		})
	}
}
```

Which results in ...

```
=== RUN   TestWildcard
=== RUN   TestWildcard/*
    prog.go:25: 
        	Error Trace:	prog.go:25
        	Error:      	Received unexpected error:
        	            	error parsing regexp: missing argument to repetition operator: `*`
        	Test:       	TestWildcard/*
=== RUN   TestWildcard/*temp
    prog.go:25: 
        	Error Trace:	prog.go:25
        	Error:      	Received unexpected error:
        	            	error parsing regexp: missing argument to repetition operator: `*`
        	Test:       	TestWildcard/*temp
=== RUN   TestWildcard/us1-*
    prog.go:28: 
        	Error Trace:	prog.go:28
        	Error:      	Not equal: 
        	            	expected: "us1-capacity"
        	            	actual  : "us1-"
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1 +1 @@
        	            	-us1-capacity
        	            	+us1-
        	Test:       	TestWildcard/us1-*
    prog.go:28: 
        	Error Trace:	prog.go:28
        	Error:      	Not equal: 
        	            	expected: "us1-load"
        	            	actual  : "us1-"
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1 +1 @@
        	            	-us1-load
        	            	+us1-
        	Test:       	TestWildcard/us1-*
=== RUN   TestWildcard/reader-?
    prog.go:28: 
        	Error Trace:	prog.go:28
        	Error:      	Not equal: 
        	            	expected: "reader-1"
        	            	actual  : "reader-"
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1 +1 @@
        	            	-reader-1
        	            	+reader-
        	Test:       	TestWildcard/reader-?
=== RUN   TestWildcard/mem?-*
    prog.go:28: 
        	Error Trace:	prog.go:28
        	Error:      	Not equal: 
        	            	expected: "mem-size"
        	            	actual  : "mem-"
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1 +1 @@
        	            	-mem-size
        	            	+mem-
        	Test:       	TestWildcard/mem?-*
    prog.go:28: 
        	Error Trace:	prog.go:28
        	Error:      	Not equal: 
        	            	expected: "mem0-cache"
        	            	actual  : "mem"
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1 +1 @@
        	            	-mem0-cache
        	            	+mem
        	Test:       	TestWildcard/mem?-*
--- FAIL: TestWildcard (0.00s)
    --- FAIL: TestWildcard/* (0.00s)
    --- FAIL: TestWildcard/*temp (0.00s)
    --- FAIL: TestWildcard/us1-* (0.00s)
    --- FAIL: TestWildcard/reader-? (0.00s)
    --- FAIL: TestWildcard/mem?-* (0.00s)
FAIL

Program exited.
```

This is because the range of a regular expression is not equivalent to a wildcard range, it requires a token to describe the allowed match set for the range (a wildcard defaults to any character).

Not matching the optional wildcard semantics from the specification means we support an additional matching option. This is allowed, though I do not think we should be supporting this currently (IIRC, there was a consideration to add regexp matching to the specification in the future). However, since we do not support wildcard matching we still need to support the catch-all `"*"` which this currently does not.

Reply:
My goal in providing this review comment is to ensure this project is compliant with the specification. This change would make us out of compliance.

> [...] authors MAY choose to support wildcard characters [...] if wildcards are not supported in general, OpenTelemetry SDKs MUST specifically recognize the single `*` wildcard as matching all instruments.

We currently do not recognize the single `*` as matching all instruments and we currently do not support wildcards.

I agree regular expressions are a more useful type for Go users, and can be a useful addition. However, we still need to be compliant with the specification regardless if we include them or not.

What do you think about unifying the name matching with something like this?

```go
type Config struct {
	instrumentName nameMatcher
	// ...
}

type nameMatcher interface {
	// Match returns if the nameMatcher matches name.
	Match(name string) bool

	// MatchMulti returns if the nameMatcher could match multiple names.
	MatchMulti() bool
}

type stringNameMatcher string

func (m stringNameMatcher) Match(name string) bool { return name == string(m) }

func (m stringNameMatcher) MatchMulti() bool { return m == "*" }

func MatchInstrumentName(name string) Option {
	return optionFunc(func(cfg Config) Config {
		cfg.instrumentName = stringNameMatcher(name)
		return cfg
	})
}

type reNameMatcher regexp.Regexp

func (m reNameMatcher) Match(name string) bool {
	re := regexp.Regexp(m)
	return re.MatchString(name)
}

func (reNameMatcher) MatchMulti() bool { return true }

func MatchInstrumentNameRegexp(re *regexp.Regexp) Option {
	return optionFunc(func(cfg Config) Config {
		cfg.instrumentName = reNameMatcher(*re)
		return cfg
	})
}
```